libs/utils/src/ballot_styles.ts

```diff
@@ -0,0 +1,128 @@
+import _ from 'lodash';
+
+import {
+  BallotStyle,
+  BallotStyleId,
+  LanguageCode,
+  Party,
+} from '@votingworks/types';
+import { Result, assertDefined, err, ok } from '@votingworks/basics';
+
+const ID_LANGUAGES_SEPARATOR = '_';
+const GROUP_ID_PARTS_SEPARATOR = '-';
+
+function generateBallotStyleGroupId(params: {
+  ballotStyleIndex: number;
+  party?: Party;
+}): BallotStyleId {
+  // const idParts = [`ballot-style`, params.ballotStyleIndex];
+  const idParts: Array<string | number> = [params.ballotStyleIndex];
+
+  if (params.party) {
+    idParts.push(params.party.abbrev);
+  }
+
+  return idParts.join(GROUP_ID_PARTS_SEPARATOR);
+}
+
+/**
+ * Generates a ballot style ID based on the given parameters in the format:
+ * `<index>[_<partyAbbreviation>]_<languageCode1>[_<languageCode2>...]`
+ */
+export function generateBallotStyleId(params: {
+  ballotStyleIndex: number;
+  languages: LanguageCode[];
+  party?: Party;
+}): BallotStyleId {
+  return [generateBallotStyleGroupId(params), ...params.languages].join(
+    ID_LANGUAGES_SEPARATOR
+  );
+}
+
+function extractBallotStyleGroupId(
+  ballotStyleId: BallotStyleId
+): BallotStyleId {
+  return ballotStyleId.split(ID_LANGUAGES_SEPARATOR)[0] || ballotStyleId;
+}
+
+/**
+ * Returns a mapping of language-agnostic ballot style "group" ID to list of
+ * (potentially language-specific) ballot styles in that group, for the given
+ * ballot style list.
+ */
+export function getBallotStyleGroups(
+  ballotStyles: readonly BallotStyle[]
+): Record<BallotStyleId, BallotStyle[]> {
+  return _.groupBy(ballotStyles, (b) => extractBallotStyleGroupId(b.id));
+}
+
+/**
+ * Returns a language-specific ballot style, for the given
+ * {@link destinationBallotStyleLanguage}, that belongs to the same
+ * language-agnostic ballot style group as the given
+ * {@link sourceBallotStyleId}.
+ */
+export function getRelatedBallotStyle(params: {
+  ballotStyles: readonly BallotStyle[];
+  destinationBallotStyleLanguage: LanguageCode;
```

Comment:
nit: I always find myself going back and forth between `destination` and `target` 😅, but in this case, I find `targetBallotStyleLanguage` to read more intuitively than `destinationBallotStyleLanguage`